[hardware_interface::RobotHW] Extend documentation of read and write methods

fjp opened the pull request:
This PR is related to [ros-controls/ros_controllers/#474](https://github.com/ros-controls/ros_controllers/pull/474) and updates/extends the documentation for `read` and `write` methods.

The PR adds a `\note` to the documentation regarding the directional perspective of these two methods. 

I've also added a [named member group](http://www.doxygen.nl/manual/grouping.html#memgroup) "Control Loop". Previously, `read` and `write` were grouped under "Hardware Interface Switching", which is not correct  I guess.

fjp committed:
doc: updates brief and detailed description of read method

fjp committed:
doc: updates docstring for write and read methods

add group name "Control Loop" to read and write methods.
Removes these methods from group Hardware Interface Switching

fjp committed:
doc: update docstring of read and write methods

fjp commented:
Using [`rosdoc_lite`](http://wiki.ros.org/rosdoc_lite) locally seems to work except that I don't know why "Hardware Interface Switching" is grouped differently than "Control Loop" and "Resource Management":

![image](https://user-images.githubusercontent.com/1286618/77653504-1191a780-6f70-11ea-8126-a2e844779159.png)

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
 
-  /**
-   * Reads data from the robot HW
+  /** \brief Used to read data from the robot hardware (represented by its interfaces) and populate the raw data.
```
Let's keep this shorter and in the same tense as the rest of the documentation. `Read data from the robot hardware`, for example.

fjp replied:
👍

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
-   * Reads data from the robot HW
+  /** \brief Used to read data from the robot hardware (represented by its interfaces) and populate the raw data.
+   *
+   * The read method is part of the control loop cycle (read, update, write) and used to populate the raw data with 
```
```suggestion
   * The read method is part of the control loop cycle (read, update, write) and is used to populate the raw data with 
```

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
+  /** \brief Used to read data from the robot hardware (represented by its interfaces) and populate the raw data.
+   *
+   * The read method is part of the control loop cycle (read, update, write) and used to populate the raw data with 
+   * the state of your robot's hardware resources (joints, sensors, actuators). In your node that handles the control loop,
```
```suggestion
   * the state of the robot's hardware resources (joints, sensors, actuators).
```

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
-   * Reads data from the robot HW
+  /** \brief Used to read data from the robot hardware (represented by its interfaces) and populate the raw data.
+   *
+   * The read method is part of the control loop cycle (read, update, write) and used to populate the raw data with 
+   * the state of your robot's hardware resources (joints, sensors, actuators). In your node that handles the control loop,
+   * you should call this method before calling controller_manager::ControllerManager::update and write().
```
I think this can be simplified, and also let's remove "your".

```suggestion
   * The read method is part of the control loop cycle (read, update, write) and is used to populate the robot state from
   * the robot's hardware resources (joints, sensors, actuators). This method should be called before controller_manager::ControllerManager::update() and write().
```

fjp replied:
👍

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
+   * the state of your robot's hardware resources (joints, sensors, actuators). In your node that handles the control loop,
+   * you should call this method before calling controller_manager::ControllerManager::update and write().
+   * 
+   * The raw data describes regions in memory of your custom robot hardware interface (subclassed from hardware_interface::RobotHW),
+   * which is usually represented by member arrays having names that give semantic meaning (pos, vel, eff) to the registered resources.
+   * The registration of memory regions (assigning pointers) is done via JointStateInterface, for read-only joints,
+   * or JointCommandInterface, for joints that accept commands and provide feedback (read and write), and its subclasses
+   * (eg. VelocityJointInterface) that make use of JointHandle or JointStateHandle through the HardwareResourceManager.
```
This seems a little out of place to me. Registering interfaces should probably be up in the documentation for `init()`, since that's where it normally happens. I don't think we need discussion about how to store the robot state in `RobotHW`, since that's pretty trivial and also very much up to the implementation.

fjp replied:
You are right, this is probably not the best place. I've put the description about interface registration here because I wasn't aware that this is usually done inside `init()`. According to the [example in the wiki](https://github.com/ros-controls/ros_control/wiki/hardware_interface#setting-up-a-new-robot) the registration is done in the constructor and I've seen other examples doing this as well, probably because they follow the wiki example. Looking at [ros_control_boilerplate](https://github.com/PickNikRobotics/ros_control_boilerplate/blob/5e40c7e875f7ed58ed061f37d4898865e382170e/src/generic_hw_interface.cpp#L66) it's done in an `init()` method. Could you please clarify where the registration of interfaces should take place or is it just personal preference?

In case it's done inside `init()`, I'll move this part of the documentation there after adjusting it accordingly.

Do you know why the `init()` method of [ros_control_boilerplate](https://github.com/PickNikRobotics/ros_control_boilerplate/blob/5e40c7e875f7ed58ed061f37d4898865e382170e/src/generic_hw_interface.cpp#L66) isn't using the same parameters as the [base class](https://github.com/ros-controls/ros_control/blob/b78bccb5f3768d1726db79acbe8837cc5187b515/hardware_interface/include/hardware_interface/robot_hw.h#L81). So far I haven't seen anyone overriding and calling [`init(ros::NodeHandle& /*root_nh*/, ros::NodeHandle &/*robot_hw_nh*/)`](http://docs.ros.org/melodic/api/hardware_interface/html/c++/classhardware__interface_1_1RobotHW.html#a39c9ede49f3219ae2f668317c55022b6) except [this test](https://github.com/ros-controls/ros_control/blob/melodic-devel/combined_robot_hw_tests/test/combined_robot_hw_test.cpp#L38). I think, if the intended place to register interfaces is in `init(ros::NodeHandle& , ros::NodeHandle &)` and not the constructor of RobotHW we should document it in the docstring of `init()` and update the wiki example. Please let me know how you think about this.

The following may be a bit off-topic regarding this review but it is related:
As far as I know `init(ros::NodeHandle& , ros::NodeHandle &)` (or just `init()` w/o parameters) - when used for registering interfaces - is called from a node's `main()` function. Wouldn't it be possible to move the call of `init(ros::NodeHandle& , ros::NodeHandle &)` inside the [controller_manager constructor](https://github.com/ros-controls/ros_control/blob/b78bccb5f3768d1726db79acbe8837cc5187b515/controller_manager/src/controller_manager.cpp#L45) to avoid having the user calling `init()`? The constructor takes the the RobotHW interface and would call the overridden `init()`.

matthew-reynolds replied:
> Could you please clarify where the registration of interfaces should take place or is it just personal preference?

Off the top of my head, I can't think of a reason it would make a difference whether it's in the constructor or `init`, I've just always tended to put it in `init` :) Unless someone else chimes in with a good reason, I don't think it matters. I also don't think there's need to recommend one way vs the other, can we word the documentation to be inclusive of either option?

> Do you know why the init() method of ros_control_boilerplate isn't using the same parameters as the base class?

That looks like a mistake to me, but let's ping @davetcoleman to know for sure. Looks like a good example of a place where `override` would help.

> Wouldn't it be possible to move the call of `init(ros::NodeHandle& , ros::NodeHandle &)` inside the controller_manager constructor to avoid having the user calling init()?

Perhaps, but that might remove a bit of flexibility in case people are doing funky custom stuff. At this point in the lifecycle of `ros_control`, changing the API without a strong reason doesn't seem like a good idea to me.

matthew-reynolds replied:
@bmagyar Can you confirm what I said in my first point above about registering interfaces in the constructor vs init?

> Off the top of my head, I can't think of a reason it would make a difference whether it's in the constructor or init, I've just always tended to put it in init :) Unless someone else chimes in with a good reason, I don't think it matters. I also don't think there's need to recommend one way vs the other, can we word the documentation to be inclusive of either option?

davetcoleman replied:
I do not recall why it is that way, probably a mistake from me or from the sources that inspired the boilerplate. Please make a PR to fix it :-)

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +152,72 @@ class RobotHW : public InterfaceManager
+   * or JointCommandInterface, for joints that accept commands and provide feedback (read and write), and its subclasses
+   * (eg. VelocityJointInterface) that make use of JointHandle or JointStateHandle through the HardwareResourceManager.
+   * 
+   * \note The name read, of this method, depends on one's directional perspective. 
+   * In the context of the robot hardware_interface::RobotHW it refers to reading state from the robot hardware into the raw data.
+   * This is opposite to the directional perspective where write() is used to fill (write to) 
+   * the raw data with the robot's resource state. Note that this is not the intended approach.
+   * 
```
I don't think we need this at all, I think your additions above explain this well enough, but if you think we need the additional clarity let's simplify it to make it even easier to understand:

```suggestion
   * \note The name 'read()' refers to reading state from the hardware.
   * This complements 'write()', which refers to writing commands to the hardware.
   * 
```

fjp replied:
I'd prefer to keep it the way you suggested and don't remove it completely. Just for extra clarity.

matthew-reynolds reviewed (changes requested):
IMO the `\brief` on `prepareSwitch` and `doSwitch` doesn't add anything, and is too long to be a brief. I think it's fine without these changes, but if you want to give it a brief, then a shorter, less detailed summary might be better than just pulling the first sentence. Something like `Check (in non-realtime) if the given controllers can be started and stopped.`, and then leave the rest of the body as it was before to add the necessary details.

I added some suggestions below, everything I wrote for `read()` can be translated to `write()` as well. The suggestions probably violate the line length limits, please ignore my bad formatting :)

Overall I think it's great but the language can be simplified a lot. In particular, I'd like to stay away from getting too close to implementation details, since we shouldn't try to force a particular style of implementation. I'm also not a fan of "raw data", so in my suggestions I used "robot state" and "command" instead.

fjp commented:
Thanks a lot for this thorough review. I'll soon take care of your suggestions.

fjp committed:
Apply suggestions from code review

Co-Authored-By: Matt Reynolds <mtreynolds@uwaterloo.ca>

fjp committed:
apply suggestions from code review

update docstrings for RobotHW class, init, read and write

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -51,8 +51,18 @@ namespace hardware_interface
+ * the names of interface types derived from \ref HardwareInterface and
  * instances of those interface types.
+ * 
+ * The registration (\ref registerInterface) of interfaces can be done in the 
+ * constructor or init() of a custom robot hardware class.
```
As mentioned in [this comment](https://github.com/ros-controls/ros_control/pull/433#discussion_r403259048) I don't know where the interface registration should take place.

matthew-reynolds replied:
I think this flexibility is good, as I mentioned I don't think it matters where it takes place.

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -51,8 +51,18 @@ namespace hardware_interface
+ * The registration (\ref registerInterface) of interfaces can be done in the 
+ * constructor or init() of a custom robot hardware class.
+ * 
+ * The robot state is described by data members of a class derived from this 
+ * base interface. The names of these data members should give semantic meaning 
+ * (pos, vel, eff) to the registered resources (joints, sensors, actuators). 
+ * Commands from controllers are represented by a member array (e.g. cmd)
+ * and populated by the controller's update method, 
+ * (controller_interface::ControllerBase::update()).
```
Not sure if this gives too much details on how to implement a derived class. If this is the wrong place, I think there should be a link to an example showing how it should be done. For example to the [wiki page](https://github.com/ros-controls/ros_control/wiki/hardware_interface#setting-up-a-new-robot). But as I've mentioned, I think this example should be updated.

matthew-reynolds replied:
IMO this is still too much detail about implementation, I don't really like recommending data types and variable names and that sort of thing, but mentioning that this class stores the state of the robot's hardware resources and outgoing commands is fine. Maybe it should live up in the first paragraph, to elaborate on "maintains a map of the hardware interfaces" (Line 49)?

fjp replied:
I agree, this is probably not the best place for data type and variable name suggestions and I'll move this part up to the class description (Line 49). 

What about updating the [example in the wiki page](https://github.com/ros-controls/ros_control/wiki/hardware_interface#setting-up-a-new-robot)? I think this could be another option: we extend the example slightly (if required) and link to it in the class description (somewhere around line 49). Another way I could imagine is to put the wiki example into this class description. This is also done in the detailed description of the [transmission_interface](http://docs.ros.org/melodic/api/transmission_interface/html/c++/classtransmission__interface_1_1TransmissionInterfaceLoader.html#details).

matthew-reynolds replied:
I like the idea of putting a minimal example (such as the wiki example you linked) in the class description, that's a lot less fragile than linking to the wiki since the wiki could always change and break the link.

fjp replied:
As mentioned below, I've put the wiki example (slightly modified) into a new `README.md` which acts as doxygen's mainpage and also as readme for the `hardware_interface` folder of the github repository.

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -70,6 +80,13 @@ class RobotHW : public InterfaceManager
 
   /** \brief The init function is called to initialize the RobotHW from a
    * non-realtime thread.
+   * 
+   * The registration of data members to interfaces is done via 
+   * \ref JointStateInterface, for read-only joints, or \ref JointCommandInterface, 
+   * for joints that accept commands and provide feedback (read and write), 
+   * including its derived classes (e.g. \ref VelocityJointInterface).
+   * These interfaces make use of \ref JointHandle or \ref JointStateHandle through 
+   * the \ref HardwareResourceManager.
```
Again, I am not sure if the registration of interfaces belongs to init or the constructor.

matthew-reynolds replied:
I think this should be reworded to be more generic - We are not limited to classes like `JointStateInterface`, `JointCommandInterface`, `JointHandle`, etc. Though these are the ones `ros_control` itself provides, users are completely free to define their own custom interfaces and handles.

fjp replied:
Good point, I didn't think about custom ones. I'll try to find a more generic description.

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,66 @@ class RobotHW : public InterfaceManager
+   * \note The name \ref read refers to reading state from the hardware.
+   * This complements \ref write, which refers to writing commands to the hardware.
+   *
+   * Querying WallTime inside \ref read is not realtime safe. The input parameters
+   * \ref time and \ref period make it possible to inject time from any 
+   * realtime backend framework.
```
I've added a reason why the `read` and `write` methods take `time` and `duration` parameters. I hope this is their intended use. Please let me know if not.

matthew-reynolds replied:
Yep, that's their intended use. Good addition.

A bit of a nitpick, but "input parameters" is a bit redundant, "parameters" is fine. I would also reword it to "inject time from a realtime source", but that might just be my personal writing preferences creeping in ;)

fjp replied:
Sounds nice :) Let's do it that way.

fjp committed:
update docstring of write method

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,66 @@ class RobotHW : public InterfaceManager
 
-  /**
-   * Writes data to the robot HW
+  /** \brief Writes commands from controllers to the robot hardware.
+   * 
+   * The write method is part of the control loop cycle (\ref read, update, \ref write) 
+   * and used to send out commands, via the hardware interfaces 
+   * (eg. JointCommandInterface and its subclasses), to the robot's hardware 
+   * resources (joints, actuators). Inside a node that handles the control loop, 
+   * this method should be called after \ref read and controller_manager::ControllerManager::update.
+   * 
+   * \note The name \ref write refers to writing (sending out) data commands,
+   * obtained from controllers, to the robot hardware resources.
+   * Use \ref read to write the state of the robot's hardware resources to data members.
```
Now this is also shorter, similar to read. Please let me know if it requires changes.

fjp commented:
> IMO the `\brief` on `prepareSwitch` and `doSwitch` doesn't add anything, and is too long to be a brief. I think it's fine without these changes, but if you want to give it a brief, then a shorter, less detailed summary might be better than just pulling the first sentence. Something like `Check (in non-realtime) if the given controllers can be started and stopped.`, and then leave the rest of the body as it was before to add the necessary details.
> 
> I added some suggestions below, everything I wrote for `read()` can be translated to `write()` as well. The suggestions probably violate the line length limits, please ignore my bad formatting :)
> 
> Overall I think it's great but the language can be simplified a lot. In particular, I'd like to stay away from getting too close to implementation details, since we shouldn't try to force a particular style of implementation. I'm also not a fan of "raw data", so in my suggestions I used "robot state" and "command" instead.

Thanks again for the review and your suggestions @matthew-reynolds. I've switched back the `\brief` onf both `prepareSwitch` and `doSwitch` and changed a lot on the other docstrings as well (not only on `read` and `write`).

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,66 @@ class RobotHW : public InterfaceManager
 
-  /**
-   * Writes data to the robot HW
+  /** \brief Writes commands from controllers to the robot hardware.
```
Let's make this present tense, and for consistency with `read()`, drop the "from controllers".

```suggestion
  /** \brief Write commands to the robot hardware.
```

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,66 @@ class RobotHW : public InterfaceManager
+   * and used to send out commands, via the hardware interfaces 
+   * (eg. JointCommandInterface and its subclasses), to your robot's hardware 
+   * esources (joints, actuators). Inside a node that handles the control loop, 
+   * this method should be called after \ref read and controller_manager::ControllerManager::update.
```
For consistency with `read()`, something like this:

```suggestion
   * The write method is part of the control loop cycle (\ref read, update, \ref write) 
   * and used to send out commands to the robot's hardware resources (joints, actuators).
   * This method should be called after \ref read and controller_manager::ControllerManager::update.
```

fjp replied:
Looks good, I'll apply this in my next commit. Unfortunately I can't do it here on GitHub as your suggestion is marked outdated.

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,66 @@ class RobotHW : public InterfaceManager
+   * esources (joints, actuators). Inside a node that handles the control loop, 
+   * this method should be called after \ref read and controller_manager::ControllerManager::update.
+   * 
+   * \note The name \ref write refers to writing (sending out) data commands,
+   * obtained from controllers, to the robot hardware resources.
+   * Use \ref read to write the state of the robot's hardware resources to data members.
```
For consistency with `read()`:

```suggestion
   * \note The name \ref write refers to writing commands to the hardware.
   * This complements \ref read, which refers to reading state from the hardware.
```

fjp committed:
Update hardware_interface/include/hardware_interface/robot_hw.h

Co-Authored-By: Matt Reynolds <mtreynolds@uwaterloo.ca>

fjp committed:
Update hardware_interface/include/hardware_interface/robot_hw.h

Co-Authored-By: Matt Reynolds <mtreynolds@uwaterloo.ca>

matthew-reynolds commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +168,65 @@ class RobotHW : public InterfaceManager
-  /**
-   * Reads data from the robot HW
+  
+  /** \name Control Loop
+   * 
+   */
+   /**\{*/
```
Should this be the same as the other groups? I'm not a Doxygen expert, just looking at consistency. Maybe this has something to do with the different grouping you see when you generate the doc? No idea :)

```suggestion
  /** \name Control Loop
   *\{*/
```

fjp replied:
Thanks for the hint! Unfortunately it still looks different, not sure how to fix this.

matthew-reynolds replied:
Oh, I think I figured it out. The `Hardware Interface Switching` group contains the `SwitchState` enum, which isn't a member func, so the group gets displayed differently. Take a look at [this page of the Doxygen manual](http://www.doxygen.nl/manual/grouping.html#memgroup). near the bottom.

Not sure whether it's better to leave the enum as is, or move it out of the group so that it's more logical in the generated HTML... Let's get other opinions.

~In either case, let's update the group definitions anyways for consistency. From the manual, it looks like the name is supposed to be in a different comment than the `\{`, but we've been getting on fine with them in the same comment, so let's stick to that.~
Edit: Whoops, should have refreshed.

fjp committed:
add suggestions of code review in robot_hw.h

fjp committed:
add README.md and use it as mainpage for doxygen

fjp commented:
I have applied most of the suggestions so far but didn't put the example from the wiki page in the class description because I think it would overload the `hardware_interface::RobotHW` class. Instead, I've added a slightly modified version of it in a newly created `README.md` which is used as doxygen's mainpage and can also be used as a readme for the `hardware_interface` folder of the github repository itself. However, if the markdown formatting doesn't work it is also possible to create a new `mainpage.dox`, move the content of `README.md` there and just add a link to the `mainpage.dox` in the `README.md` and the delete the rest. One thing that is missing now is the `\htmlinclude manifest.html` which is usually found in a `mainpage.dox` of a ROS package. Note sure if this can be added inside the README.md. Please let me know what you think, any suggestions are welcome.

fjp commented on hardware_interface/README.md:
```diff
@@ -0,0 +1,299 @@
+  }
+
+private:
+  hardware_interface::JointStateInterface jnt_state_interface;
+  hardware_interface::PositionJointInterface jnt_pos_interface;
```
Why exactly are both needed? Isn't [PositionJointInerface](http://docs.ros.org/melodic/api/hardware_interface/html/c++/classhardware__interface_1_1PositionJointInterface.html) sufficent because it inherits from [JointCommandInterface](http://docs.ros.org/melodic/api/hardware_interface/html/c++/classhardware__interface_1_1JointCommandInterface.html) which can be used for reading and writing? Or are both required because all [existing controllers](https://github.com/ros-controls/ros_controllers) require a `JointStateInterface` to read the state of the robot's resources and don't make use of read/write interfaces (only custom controllers could)?

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -80,6 +107,7 @@ class RobotHW : public InterfaceManager
    */
   virtual bool init(ros::NodeHandle& /*root_nh*/, ros::NodeHandle &/*robot_hw_nh*/) {return true;}
```
Is there a reason why all parameters are commented in the headers? I know only the type is required in the declaration of a method but I am asking because doxygen gave me warnings of the missing parameter names.

matthew-reynolds replied:
That's... A good question. I suspect that's a mistake. Commenting out unused parameters at the function definition is sometimes done to suppress compiler warnings, but I don't think there's any reason to comment them out in the declaration like this. Would you mind un-commenting them here in the header, and double-checking that this satisfies both Doxygen and the compiler? (ie there are no compiler warnings when you compile with `-Wunused-parameter`/`-Wunused`/`-Wextra`)

matthew-reynolds replied:
See #399. Looks like these were probably extra accidental changes.

fjp replied:
Thanks for the link. I've uncommented the parameters and compiled the `hardware_interface` tests (where `robot_hw.h` gets included) using: `catkin build -v --pre-clean --cmake-args -DCMAKE_C_FLAGS="-Wall -Wextra -Wunused-parameter -Wunused" --catkin-make-args tests` without any warnings. Should I commit these changes (uncommenting parameters) to this PR or in a separate one?

matthew-reynolds replied:
Let's split it out, this is already going to be a fairly lengthy PR. Thanks :)

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -148,33 +178,59 @@ class RobotHW : public InterfaceManager
     ERROR
   };
 
-  // Return (in realtime) the state of the last doSwitch()
+  /** \brief Return (in realtime) the state of the last doSwitch() */
   virtual SwitchState switchResult() const
   {
     return DONE;
   }
 
-  // Return (in realtime) the state of the last doSwitch() for a given controller
+  /** \brief Return (in realtime) the state of the last doSwitch() for a given controller */
   virtual SwitchState switchResult(const ControllerInfo& /*controller*/) const
   {
     return DONE;
   }
+  /**\}*/
 
-  /**
-   * Reads data from the robot HW
+  /** \name Control Loop
+   *\{*/
```
I don't know how to make these groups look alike. `Hardware Interface Switching` seems to be more important to doxygen than `Resource Management` and `Control Loop`.

matthew-reynolds replied:
See https://github.com/ros-controls/ros_control/pull/433#discussion_r405916399

fjp commented on hardware_interface/rosdoc.yaml:
```diff
@@ -1,5 +1,5 @@
   name: C++ API
   output_dir: c++
-  file_patterns: '*.c *.cpp *.h *.cc *.hh *.dox'
+  file_patterns: '*.c *.cpp *.h *.cc *.hh *.dox *.md'
```
This is required to use a mardown file (`README.md`) as doxygen's mainpage. See also [`rosdoc_lite`](http://wiki.ros.org/Doxygen#Embedding_Markdown_Pages) how to include markdown files. Only "problem" is that the `\htmlinclude manifest.html` is missing, which is usually present in default `mainpage.dox` of ROS packages.

fjp commented on hardware_interface/README.md:
```diff
@@ -0,0 +1,299 @@
+A list of available hardware interfaces (provided via the HardwareResourceManager) 
+as of this writing:
+
+- [JointCommandInterface](\ref hardware_interface::JointCommandInterface): 
```
Not sure if these links will work in the README.md when viewed on the GitHub repo. It is working though in the generated doxygen documentation.

matthew-reynolds replied:
Looks like they do not work :(

https://github.com/fjp/ros_control/tree/melodic-devel/hardware_interface

fjp commented on hardware_interface/include/hardware_interface/robot_hw.h:
```diff
@@ -51,8 +51,26 @@ namespace hardware_interface
+ * should provide semantic meaning (e.g. pos, vel, eff, cmd). 
+ * Commands from controllers are populated by the controller's update method, 
+ * (controller_interface::ControllerBase::update()).
+ * For each resource a \ref JointStateHandle (for read only joints), 
+ * \ref JointHandle (for read and write joints) or custom handle can be used 
```
This question is similar to the [previous one](https://github.com/ros-controls/ros_control/pull/433/files#r405166152). Examples I've seen so far use both (`JointStateHandle` and `JointHandle`) but wouldn't `JointHandle` be enough as it can be used for reading joint state and writing commands?

fjp commented on hardware_interface/README.md:
```diff
@@ -0,0 +1,299 @@
+        const ros::Duration period = time - prev_time;
+        
+        // Execution of the actual control loop.
+        robot.read();
+        cm.update(time, period);
+        root.write();
```
Where exactly in the control loop should we use the `transmission_interface` if a robot's transmissions need to be defined in software? Is this done inside `read()` and `write()` methods? Or is this already handled in the URDF definition (at least for the available transmissions such as [`SimpleTransmission`](http://docs.ros.org/melodic/api/transmission_interface/html/c++/classtransmission__interface_1_1SimpleTransmission.html)?

fjp commented on hardware_interface/README.md:
```diff
@@ -0,0 +1,299 @@
+public:
+  MyRobot() 
+  { 
+    // Initialization of the robot's resources (joints, sensors, actuators) and
+    // interfacs can be done here or inside init().
```
Maybe we could add here that it is possible to get the joint names from the URDF via the parameter server like the [ros_control_boilerplate/generic_hw_interface](https://github.com/PickNikRobotics/ros_control_boilerplate/blob/5e40c7e875f7ed58ed061f37d4898865e382170e/src/generic_hw_interface.cpp#L52) seems to do it (using [rosparam_shortcuts](https://github.com/PickNikRobotics/rosparam_shortcuts)).

matthew-reynolds commented:
Since PR this is growing larger, I just want to ping @ros-controls/ros_control for more opinions.

@fjp I will need a few days to read your newest changes :sweat_smile:

matthew-reynolds commented:
@fjp Sorry to go back on you here, but I think this PR is losing focus a bit. Do you mind leaving the markdown group fixes, `\brief`s, and the `read()` and `write()` descriptions, but splitting the rest of this out into a new PR? I think those other changes need more discussion, particularly the inclusion of the big README.

The few changes I listed above are basically ready to be merged and address specific issues rather than general improvement of the documentation.

fjp commented:
@matthew-reynolds no worries, splitting this up is a good idea. I'll leave the changes you mention and move the large changes of the documentation (README) to a new PR.

fjp commented:
@matthew-reynolds I've reverted the large changes (remove README, revert rosdoc.yaml, remove details docstrings of RobotHW class and init method) of this PR in a new branch [extend-read-write-docstrings](https://github.com/fjp/ros_control/tree/extend-read-write-docstrings) in my fork. In this branch I've also squashed most of the commits as the commit messages were similar. Should I use this new branch for a new PR? 

Or do you prefer using this PR with one additional commit to `melodic-devel` branch that removes the README.md, reverts rosdoc.yaml and removes the added details of the RobotHW class description and init method?

matthew-reynolds commented:
I don't think it's a big deal either way. On one hand, if we stay here, we preserve the discussion. But on the other hand, it's getting hard to navigate this PR since there's so much discussion and back-and-forth. Plus we can (And should) link to this PR in the new one if we go that route.

Not sure if the higher-ups have an opinion one way or the other, but I'd say probably doesn't matter, up to you.

fjp commented:
Thank for the feedback @matthew-reynolds. As this PR has quite a lot of comments and is rather confusing I'd like to make a new PR on a new branch ([extend-read-write-docstrings](https://github.com/fjp/ros_control/tree/extend-read-write-docstrings)). Also I think it is cleaner if I start to work on branches and not on `melodic-devel`, even if it's a fork. 

Therefore, I created a new PR #444 which mainly takes care of the `read` and `write` method docstrings and the doxygen group names. For the rest of the changes in this PR (extend documentation of `hardware_interface`) I'll come up with a new PR once #444 get's sorted out. I hope this is ok?

bmagyar commented:
I have merged #444 and also picked it for Noetic. I'd recommend opening a new PR with 100-200 lines of unchallenged changes from this PR so we can guarantee a somewhat smooth review process.

fjp commented:
Great, thanks @bmagyar. I'll do my best to keep the changes in the upcoming PR small.

fjp commented:
This PR should become obsolete once #457 is merged.

bmagyar commented:
Closing on favour of #457 , thanks guys for the thorough discussion here
